### PR TITLE
Blocks: Add support for Collections

### PR DESCRIPTION
**Pull request by kraftbj**
Supports #14598
Replaces and closes #14721

In Gutenberg 7.3, the block editor now supports "Collections". In the block picker, a Collection looks like a category, except blocks can also have a category.

The use case here is when do people think "I need a business hours block. I bet that's in the Jetpack category!"... probably not. They probably think of the function -- an embed, a widget, etc.

The problem is a block can only be in one category so devs have generally grouped them by the block provider.

Collections allow the block provider (Jetpack in our case) present all of our blocks together—like we have done with a custom category—while also listing it in a context-appropriate category.

For those running Gutenberg 7.3 or WordPress 5.4:
2020-01-23 at 9 14 PM

For those not running Gutenberg 7.3 or on WordPress 5.3, it retains the original behavior:
2020-01-23 at 9 15 PM

#### Changes proposed in this Pull Request:
* Adds a 'Jetpack' block collection.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* n/a

#### Testing instructions:
* On a connected Jetpack site WITHOUT Gutenberg, add a new post. Verify all the blocks are in a Jetpack category when viewing the block inserter.
* On a connected Jetpack site WITH Gutenberg, add a new post. With this patch, the Business Hours block will be in a Jetpack collection (looks just like a category) AND the widgets category.

AS OF NOW, when running G 7.3, the rest of Jetpack's blocks will fail since we're not adding them to a `Jetpack` category. We'll need to update the other blocks before landing it, but opening the PR now to get opinions on the implementation.

I opted to need to update blocks since if we allow those blocks with other categories to be added to a collection and those with the 'legacy' Jetpack category, the inserter will have two "Jetpack" entries -- the category and the collection -- so that's not really an option. 

#### Proposed changelog entry for your changes:
* Block Editor: Blocks are now listed both as part of a Jetpack Collection and in their proper category. (Requires Gutenberg 7.3)

### CODE REVIEW

**Comment by scruffian** [comment]
This might make sense to be done in tandem with this PR: https://github.com/WordPress/gutenberg/pull/19279

Since the categories are all set to change

**Comment by simison** [comment]
When merged, we should confirm that re-organizing categories on wpcom is still functional: https://github.com/Automattic/wp-calypso/pull/37057

File: extensions/blocks/business-hours/index.js
Comment (simison) [inline extensions/blocks/business-hours/index.js]: I would centralize `@wordpress/blocks` import and typeof check into a helper function at `extensions/shared/block-category.js` and do `supportsCollections()` or similar.

**Comment by simison** [comment]
Would be nice to get this in before code freeze on Monday evening in time for WP 5.4 release which supports thi feature. @lessbloat

**Comment by mtias** [comment]
Perhaps:

- Calendly → Widgets
- Contact Info → Formatting
- Eventbrite Checkout → Widgets
- Form → Widgets
- GIF → Formatting
- Google Calendar → Embeds
- Map → Common
- Markdown → Formatting
- Open Table → Widgets
- Pinterest → Embeds
- Repeat Visitor → Widget
- Revue → Widget
- Slideshow → Common
- Star Rating → Formatting
- Tiled Gallery → Common

Register new category: _Earn_ or _Premium_.

- Ads
- Simple Payments
- Recurring Payments

**Comment by kraftbj** [comment]
Rebased and not finished yet, but need to log off.

What categories for these blocks:
- Amazon
- Business Hours (Widgets in this PR right now)
- Instagram
- Mailchimp
- Podcast Player
- Related Posts
- Subscriptions

Still need to add the new category for Ads, Simple, Recurring Payments.

**Comment by apeatling** [comment]
@kraftbj Let me know if you need any help getting this one finished off.

**Comment by obenland** [comment]
#14598 has [a spreadsheet with suggested categories](https://docs.google.com/spreadsheets/d/1Oua0oDUhGJJtabatoBS3XBzZ6K3oUiB5-kSdWl00M-I/edit#gid=0) based on the discussion in p58i-8yl-p2. #14721 is the corresponding PR. It does away with the `'jetpack'` category regardless of whether collections are available or not, as suggested in that post. Should we do the same here?

**Comment by kraftbj** [comment]
>  It does away with the 'jetpack' category regardless of whether collections are available or not, as suggested in that post. Should we do the same here?

I'm hesitant to do that since documentation notes you can find the JP blocks under "Jetpack" and existing users who aren't updated Core would lose that flow. When 5.4 is Jetpack's minimum, we can remove the fallback.

Thanks for the link to the spreadsheet.

**Comment by kraftbj** [comment]
Updated the PR with the two new categories and updating all blocks with the spreadsheet.

OpenTable struck me as odd being "Earn" instead of "Marketing" or "Embed, but included it as is.

Pinterest also was the only "social" one and there isn't that category yet. I made it "embed" instead for now.

Podcast Player wasn't in the spreadsheet. Marked it as an embed.

Rating Star wasn't in the spreadsheet. Deferred to the suggestion in the comment above.

Related Posts is listed as the only one in a non-existent "blog" category. Marked it as Embed.

**Comment by kraftbj** [comment]
Noting that the wp.com patch needs a little massaging for the maps block. Nothing major, but I'll wait until this has landed to work on it so it isn't overridden with future work in this PR.

**Comment by obenland** [comment]
Pinging @davemart-in for review as this fixes 80% of #14598 for Ajax.

**Review by jeherve — CHANGES_REQUESTED** [review]
This looks good. However, in WP 5.4, clicking on the Jetpack category while another Jetpack collection is open triggers a JavaScript error and breaks the editor, maybe because we're trying the same block multiple times.

```
Uncaught TypeError: Cannot read property 'attributes' of undefined
    at createBlock (blocks.js?ver=423068d7079f57cf9f02458ccb4a6131:7448)
    at getBlockFromExample (blocks.js?ver=423068d7079f57cf9f02458ccb4a6131:7874)
    at blocks.js?ver=423068d7079f57cf9f02458ccb4a6131:7875
    at arrayMap (lodash.js?ver=4.17.15:639)
    at map (lodash.js?ver=4.17.15:9554)
    at getBlockFromExample (blocks.js?ver=423068d7079f57cf9f02458ccb4a6131:7874)
    at InserterMenu.render (dmp.js:19)
    at finishClassComponent (react-dom.js?ver=16.9.0:17177)
    at updateClassComponent (react-dom.js?ver=16.9.0:17132)
    at beginWork$1 (react-dom.js?ver=16.9.0:18643)
```

File: extensions/shared/block-category.js
Comment (kraftbj) [reply extensions/shared/block-category.js:45]: Good call. 79f97f037

**Comment by simison** [comment]
>It does away with the 'jetpack' category regardless of whether collections are available or not, as suggested in that post. Should we do the same here?

In the plugin, the Jetpack category is also a great way to communicate value Jetpack is giving without hiding the blocks only into different categories.

On .com however that's not so important and the Jetpack category can be confusing. It could make sense to hide Jetpack for .com but leave it for Jetpack the plugin.

[There are helpers](https://github.com/Automattic/jetpack/blob/fa68e186d52784d3a83b6e27c1e06a73b295654d/extensions/shared/site-type-utils.js) for that in case you need.

**Comment by davemart-in** [comment]
I'm seeing a fatal error when just clicking around. See: https://a.cl.ly/bLumEllP This is using a Jurassic Ninja instance.

**Comment by kraftbj** [comment]
I haven't been able to duplicate that error yet, but I've been using the inline inserter. Still looking.

> On .com however that's not so important and the Jetpack collection in block picker can be confusing. It could make sense to hide Jetpack collection in .com, but leave it for the plugin.

We can do that in a follow-up PR IMO.

**Comment by simison** [comment]
>We can do that in a follow-up PR IMO.

Absolutely 👍 Wasn't implying it should be done here. :-)

**Comment by kraftbj** [comment]
@davemart-in Can you check it out again?

The contact form's child blocks weren't assigned to a new category yet (and Jetpack wasn't being registered anymore). The inline picker didn't have a problem, but the top inserter was trying to load a dynamic preview using blocks that didn't exist -- thus fataling. Fixed in 74f11dd.

**Comment by davemart-in** [comment]
@kraftbj I'm still seeing the error unfortunately. I'm seeing [this under the JP beta section](https://a.cl.ly/JruW25qq). I assume that is the latest.

One other oddity: I'm not seeing simple payments in the earn collection or the Jetpack collection: https://a.cl.ly/mXuqezq7

**Comment by kraftbj** [comment]
Click Update on the top. The branch has updated but your site isn't running it. 

I'll check out Simple Payments.

**Comment by jeherve** [comment]
> I'm seeing this under the JP beta section. I assume that is the latest.

You'd have to click on Update at the top of the page to get the latest version.

> I'm not seeing simple payments in the earn collection or the Jetpack collection

That's to be expected as the block is only available to folks with a plan. You would have to purchase a plan, or enable block nudges like so:
`add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_true' );`

**Review by jeherve — DISMISSED** [review]
I am not seeing the error anymore. It seems to work well!

I wonder if we should really be inserting our blocks before core blocks in core categories like Embeds?

image

**Comment by davemart-in** [comment]
Confirmed. That bug is fixed now @kraftbj. 👍 Thanks!

**Comment by simison** [comment]
@jeherve there is separate script on wpcom that reorganises categories in case we want to do that only there: https://github.com/Automattic/wp-calypso/blob/master/apps/wpcom-block-editor/src/wpcom/features/reorder-block-categories.js

File: extensions/shared/block-category.js
Comment (jeherve) [inline extensions/shared/block-category.js:46]:
```suggestion
		title: __( 'Earn', 'jetpack' ),
```

File: extensions/shared/block-category.js
Comment (jeherve) [inline extensions/shared/block-category.js:56]:
```suggestion
		title: __( 'Grow', 'jetpack' ),
```

**Review by jeherve — CHANGES_REQUESTED** [review]
Let's add the textdomains here.

**Review by jeherve — APPROVED** [review]
This looks good to me now. I'd suggest merging this as is for the 8.4 release, and work on improving these collections in a future PR:
- Handle wpcom differently
- Have Jetpack blocks come after core blocks in core categories.

**Comment by jeherve** [comment]
> Handle wpcom differently

To close the loop on this: #15310 

> Have Jetpack blocks come after core blocks in core categories.

I have been thinking about this some more, and I'm honestly not quite sure what we should aim for. Should Jetpack blocks be available before Core blocks in categories like Embeds?

image

My first instinct was to say no, our blocks should come after Core's. Core's blocks would most likely be used more often; they were added to Core more often. On the other hand I can see why having our own blocks first should give us some extra visibility.

Furthermore, Core currently adds custom blocks before core blocks in core categories, which seems to indicate that they're okay wit this?

**Comment by davemart-in** [comment]
> My first instinct was to say no, our blocks should come after Core's. Core's blocks would most likely be used more often; they were added to Core more often. On the other hand I can see why having our own blocks first should give us some extra visibility.

Could they just be listed alphabetically?

**Comment by jeherve** [comment]
> Could they just be listed alphabetically?

I suppose it could be another option. Would we then change all block ordering to alphabetical as soon as Jetpack is active, or only in categories that include Jetpack blocks?

**Comment by kraftbj** [comment]
Cherry-picked to `branch-8.4` via 9c91c106a .

**Comment by kraftbj** [comment]
r205423-wpcom

**Pull request merged**
